File: backend/src/utilidades/youtube.py
from urllib.parse import parse_qs, urlparse

YOUTUBE_EMBED_PARAMS = "autoplay=1&mute=1&controls=0&rel=0&modestbranding=1&playsinline=1"

def extraer_youtube_id(referencia: str | None) -> str | None:
    if not referencia:
        return None

    valor = referencia.strip()
    if len(valor) == 11 and all(c.isalnum() or c in "_-" for c in valor):
        return valor

    parsed = urlparse(valor)
    host = parsed.netloc.lower()
    path = parsed.path.strip("/")

    if "youtu.be" in host and path:
        return path.split("/")[0][:11]

    if "youtube.com" in host or "youtube-nocookie.com" in host:
        if path == "watch":
            return parse_qs(parsed.query).get("v", [None])[0]
        if path.startswith("embed/"):
            return path.split("/", 1)[1][:11]
        if path.startswith("shorts/"):
            return path.split("/", 1)[1][:11]

    return None

def construir_url_embed_youtube(referencia: str | None) -> str | None:
    video_id = extraer_youtube_id(referencia)
    if not video_id:
        return None
    return (
        f"https://www.youtube-nocookie.com/embed/{video_id}?{YOUTUBE_EMBED_PARAMS}"
        f"&loop=1&playlist={video_id}&enablejsapi=1"
    )

File: frontend/src/services/youtube.ts
const DEFAULT_PARAMS = "autoplay=1&mute=1&controls=0&rel=0&modestbranding=1&playsinline=1";

function getOriginParam(): string {
  if (typeof window === 'undefined') return '';
  try {
    return `&origin=${encodeURIComponent(window.location.origin)}`;
  } catch {
    return '';
  }
}

export function extractYouTubeId(reference: string | null | undefined): string | null {
  if (!reference) return null;

  const value = reference.trim();
  if (/^[A-Za-z0-9_-]{11}$/.test(value)) {
    return value;
  }

  try {
    const url = new URL(value);
    const host = url.hostname.toLowerCase();
    const path = url.pathname.replace(/^\/+/, '');

    if (host.includes('youtu.be') && path) {
      return path.split('/')[0].slice(0, 11);
    }

    if (host.includes('youtube.com') || host.includes('youtube-nocookie.com')) {
      if (path === 'watch') {
        const v = url.searchParams.get('v');
        return v ? v.slice(0, 11) : null;
      }
      if (path.startsWith('embed/')) {
        return path.split('/')[1]?.slice(0, 11) || null;
      }
      if (path.startsWith('shorts/')) {
        return path.split('/')[1]?.slice(0, 11) || null;
      }
    }
  } catch {
    return null;
  }

  return null;
}

export function toYouTubeEmbedUrl(reference: string | null | undefined): string | null {
  const videoId = extractYouTubeId(reference);
  if (!videoId) return null;
  const originParam = getOriginParam();
  return (
    `https://www.youtube-nocookie.com/embed/${videoId}?${DEFAULT_PARAMS}` +
    `&loop=1&playlist=${videoId}&enablejsapi=1${originParam}`
  );
}

export function toYouTubePlaylistEmbedUrl(
  references: string[],
  startIndex = 0
): string | null {
  const ids = references
    .map((reference) => extractYouTubeId(reference))
    .filter((id): id is string => Boolean(id));

  if (ids.length === 0) return null;

  const normalizedIndex = ((startIndex % ids.length) + ids.length) % ids.length;
  const ordered = [...ids.slice(normalizedIndex), ...ids.slice(0, normalizedIndex)];
  const first = ordered[0];
  const playlist = ordered.join(',');
  const originParam = getOriginParam();

  return (
    `https://www.youtube-nocookie.com/embed/${first}?${DEFAULT_PARAMS}` +
    `&loop=1&playlist=${playlist}&enablejsapi=1${originParam}`
  );
}

export function toYouTubeThumbnailUrl(reference: string | null | undefined): string | null {
  const videoId = extractYouTubeId(reference);
  if (!videoId) return null;
  return `https://i.ytimg.com/vi/${videoId}/mqdefault.jpg`;
}
